### ISSUE TITLE
[BUG] Dask shuts down workers when host RAM is about to run out

### ISSUE THREAD

**Issue by rommelDB** [BlazingDB/blazingsql#736]
**Describe the bug**
The engine is able to calculate when it is going to start moving frames from GPU memory to Host memory and then to disk. However, with very large data queries it is not working well since the workers are being shut down with the message below:

distributed.nanny - WARNING - Worker exceeded 95% memory budget. Restarting

**Steps/Code to reproduce bug**
Run a query with large data enough to use the disk after run out the host memory.

**Expected behavior**
It should work smoothly with the bigger than GPU architecture.

**Environment overview**
 - Environment location: Bare-metal
 - Method of cuDF install: conda
 - Method of blaazingsql install: from source, branch-0.14

**Additional context**
Apparently the engine should be considering this Dask feature to shutdown workers automatically (by default the threshold is set to 0.95).

https://github.com/dask/distributed/blob/d8f46d502a561bd86cc5c08eec0fd906a4ba700d/distributed/distributed.yaml#L80-L86

**Comment by rommelDB** [comment]
I ran different experiments investigating this issue. First, I tried to replicate the original issue locally, by varying the memory limit assigned to each Dask worker. When workers start with a low memory limit (i.e. 2GB), in most cases, such workers pause, and the query execution hangs. I would understand that this behavior is due to the sudden requirement of large amount of memory.

Then, I tried to decrease the GPU memory, and host memory limits thresholds on the engine. From here, we can see different types of crashes. So, what I understand is with BiggerThanGPU architecture, the engine can perform batch loads and process such data within the preconfigured memory limits. However, some CUDF operators, such as joins, end up requiring device memory that exceeds the available memory at any given time, so then, the program just crashes with diverse memory error messages.

I am aware that adjusting the value of JOIN_PARTITION_SIZE_THRESHOLD, some of these problems are fixed in the case of joins. Here, I wonder if this value could be calculated automatically, adapting itself to the constraints of the target hardware environment.

On the other hand, maybe it is feasible to catch some memory exceptions thrown by CUDF, when it requires more memory than is available. So, if there are memory batches on the device that are not being used for that operation, they could temporarily be transferred to the host memory; finally, we could try to rerun that operation.
